boot/assets.go

```diff
@@ -50,19 +151,47 @@ type TrustedAssetsInstallObserver struct {
 //
 // Implements gadget.ContentObserver.
 func (o *TrustedAssetsInstallObserver) Observe(op gadget.ContentOperation, affectedStruct *gadget.LaidOutStructure, root, realSource, relativeTarget string) (bool, error) {
-	// TODO:UC20:
-	// steps on write action:
-	// - copy new asset to assets cache
-	// - update modeeenv
+	if affectedStruct.Role != gadget.SystemBoot {
+		// only care about system-boot
+		return true, nil
+	}
+
+	if o.bl == nil {
+		bl, err := bootloader.ForGadget(o.gadgetDir, root, &bootloader.Options{NoSlashBoot: true})
+		if err != nil {
+			return false, fmt.Errorf("cannot find bootloader: %v", err)
+		}
+		tbl, ok := bl.(bootloader.TrustedAssetsBootloader)
+		if !ok {
+			return true, nil
+		}
+		trustedAssets, err := tbl.TrustedAssets()
+		if err != nil {
+			return false, fmt.Errorf("cannot list %q bootloader trusted assets: %v", bl.Name(), err)
+		}
+		o.trustedAssets = trustedAssets
+		o.bl = bl
+	}
+	if len(o.trustedAssets) == 0 || !strutil.ListContains(o.trustedAssets, relativeTarget) {
+		// not one of the trusted assets
+		return true, nil
+	}
+	ta, added, err := o.cache.Add(realSource, o.bl.Name(), filepath.Base(relativeTarget))
+	if err != nil {
+		return false, err
+	}
+	if added {
```

Comment:
It doesn't matter for install because the cache goes to the ephemeral root, but this merits a comment or a change. In general we should add the trackedAssets to the list based whether is in the list or not, not based if it was added to the cache or not.

Reply:
I think the point I'm trying to make is that the added flag from Add is probably dangerous in general.